bliss/models/galaxy_net.py

```diff
@@ -1,174 +1,262 @@
-        recon_mean = self(images, background)
-        residuals = (images - recon_mean) / torch.sqrt(recon_mean)
-        loss = self.get_loss(images, recon_mean)
+        recon_mean_main = self._main_forward(images, background)
+        loss_main = self._get_likelihood_loss(images, recon_mean_main)
+        self.log("val/loss_main", loss_main)
 
-        # metrics
+        recon_mean_residual = self._residual_forward(images - recon_mean_main)
+        loss_residual = self._get_residual_model_loss(images, recon_mean_main, recon_mean_residual)
+        self.log("val/loss_residual", loss_residual)
+
+        recon_mean_final = F.relu(recon_mean_main + recon_mean_residual)
+        loss = self._get_likelihood_loss(images, recon_mean_final)
         self.log("val/loss", loss)
+
+        # metrics
+        residuals = (images - recon_mean_final) / torch.sqrt(recon_mean_final)
+        residuals_main = (images - recon_mean_main) / torch.sqrt(recon_mean_main)
+        recon_mean_residual = recon_mean_residual / torch.sqrt(recon_mean_main)
         self.log("val/max_residual", residuals.abs().max())
-        return {"images": images, "recon_mean": recon_mean, "residuals": residuals}
+        return {
+            "images": images,
+            "recon_mean_main": recon_mean_main,
+            "recon_mean_residual": recon_mean_residual,
+            "recon_mean": recon_mean_final,
+            "residuals": residuals,
+            "residuals_main": residuals_main,
+        }
 
     def validation_epoch_end(self, outputs):
         """Validation epoch end (pytorch lightning)."""
 
-        # combine all images and recon_mean's into a single tensor
-        images = torch.cat([output["images"] for output in outputs])
-        recon_mean = torch.cat([output["recon_mean"] for output in outputs])
-        residuals = torch.cat([output["residuals"] for output in outputs])
+        output_tensors = {
+            label: torch.cat([output[label] for output in outputs]) for label in outputs[0]
+        }
+
+        fig_random = self._plot_reconstruction(output_tensors, mode="random")
+        fig_worst = self._plot_reconstruction(output_tensors, mode="worst")
+        grid_example = self._plot_grid_examples(output_tensors)
 
-        fig_random = self.plot_reconstruction(images, recon_mean, residuals, mode="random")
-        fig_worst = self.plot_reconstruction(images, recon_mean, residuals, mode="worst")
         if self.logger:
-            self.logger.experiment.add_figure(f"Random Images {self.current_epoch}", fig_random)
-            self.logger.experiment.add_figure(f"Worst Images {self.current_epoch}", fig_worst)
+            self.logger.experiment.add_figure(
+                f"Epoch:{self.current_epoch}/Random Images", fig_random
+            )
+            self.logger.experiment.add_figure(f"Epoch:{self.current_epoch}/Worst Images", fig_worst)
+            self.logger.experiment.add_figure(
```

Comment:
Maybe split declaring the title and the `add_figure` command into two lines to avoid multi-line commands?